### ISSUE TITLE
`let b:dispatch = '-compiler=<compiler>'` not working as expected

### ISSUE THREAD

**Issue by zealotrush** [tpope/vim-dispatch#315]
I want to do C code linting in Vim. First, I created a compiler plugin as shown below.

```vim
" ~/.vim/after/compiler/clang.vim
CompilerSet makeprg=clang\ -fsyntax-only\ %:S
```

Then, to test the plugin, I wrote a simple hello world C program, with a deliberate mistake in it.

```c
// main.c
#include <stdio.h>
int main(void)
{
    printf("hello world\n")   // <-- missing semicolon!
    return 0;
}
```

If I run the following dispatch command directly in the command-line window,

```vim
:Dispatch -compiler=clang
```

I can get the expected result in the quickfix list.

```
main.c|4 col 28| error: expected ';' after expression
||     printf("hello world\n")     // <-- missing semicolon!
||                            ^
||                            ;
|| 1 error generated.
```

However, if I use `b:dispatch`, like the following:

```
:let b:dispatch = '-compiler=clang'
```

then run `:Dispatch`, I only get a message:

```
Success:! (job/85918)
```

which is not the expected result. Where did I get it wrong?

**Comment by tpope** [comment]
You're in the realm of undefined behavior. Both of those should probably produce an error message if we were being strict about it. You want `CompilerSet makeprg=clang` and `let b:dispatch = 'clang -fsyntax-only %:S'`, or perhaps `let b:dispatch = '-compiler=clang clang -fsyntax-only %:S'` if you don't want to leave the compiler plugin selection up to chance.

**Comment by zealotrush** [comment]
I'm wondering what shell command was run exactly by the `:Dispatch` command described above. Clearly, it was not the command I expected.

If it was a bare `clang` command without any arguments, it should produce the following error:

```
|| clang: error: no input files
```

Or, if it was the default value `make`, the result should be:

```
|| make: *** No targets specified and no makefile found.  Stop.
```

Is there any way for me to check that out?

**Comment by tpope** [comment]
Looks like it's running nothing at all. An empty string. `$SHELL -c ''`.

**Comment by tpope** [comment]
I've tweaked it to be a bit more intelligent, but I still strongly believe that `%:S` belongs in `b:dispatch`, *not* the compiler plugin. Hard coding arguments like that greatly limits versatility.

**Comment by tpope** [comment]
If you want to do it with `.projections.json`, I recommend sucking it up and repeating yourself a bit:

```json
{
  "my/target/source/*.c": {
    "type": "source",
    "path": [
      "some/path",
      "other/path"
    ],
    "dispatch": "clang -fsyntax-only -Wall -Wextra -Isome/path -Iother/path {file}"
  }
}
```

If you're doing this across enough projects that the repetition becomes a significant pain point, look into setting `b:dispatch` on the `autocmd User ProjectionistActivate` event. You can start with the value that Fugitive sets for you and then append your `-I` values to it.

**Comment by tpope** [comment]
Dispatch does have an undocumented way to do this for `:Make` too, using a special sequence in `'errorformat'`:

```vim
" ~/.vim/after/compiler/clang.vim
CompilerSet errorformat+=%\\&default=-fsyntax-only\ -Wall\ -Wextra\ %f
```

This lets `:Make` call `clang -fsyntax-only -Wall -Wextra current_filename.c` while `:Make foo.c` will call `clang foo.c`, best of both worlds. Pair this with a change to `'makeprg'` to add the `-I` variables in `autocmd User ProjectionistActivate` and you should have everything you need.

I only recommend using this trick when the command in question is useless without an argument. (That is, I wouldn't use it with a command like `make` because then there's no good way to call the default make task).

**Comment by zealotrush** [comment]
Thank you for your help!

I tried `autocmd User ProjectionistActivate` and it worked surprisingly well. I kept the `~/.vim/after/compiler/clang.vim` file unchanged, which constructs the full `'makeprg'` command (including the `%:S` part, ugly but works). Then I add the following script:

```vim
" ~/.vim/plugin/projectionist-config.vim
" loading guard
augroup auto_compiler_for_c_cpp
  autocmd!
  autocmd FileType c,cpp autocmd User ProjectionistActivate compiler clang
augroup END
```

Now the `:compiler clang` was called at the right moment, and the `-I<path>` part is configured correctly.

I didn't try the undocumented special `errorformat`, because it wasn't obvious to me how to configure the compiler program depend on `'filetype'`, e.g., `clang` for C, and `clang++` for C++. Plus, I don't see much difference between setting these things in `'makeprg'` and in `'errorformat'`, after all, they all involve the `%:S`/`%f` part.
